src/main/java/software/amazon/cloudformation/resource/Validator.java

```diff
@@ -68,7 +95,72 @@ public void validateObject(final JSONObject modelObject, final JSONObject defini
      * @throws ValidationException Thrown for any schema validation errors
      */
     public void validateResourceDefinition(final JSONObject definition) throws ValidationException {
+        // inject/replace $schema URI to ensure that provider definition schema is used
+        definition.put("$schema", RESOURCE_DEFINITION_SCHEMA_URI.toString());
         validateObject(definition, definitionSchemaJsonObject);
+        // validateObject cannot validate schema-specific attributes. For example if definition
+        // contains "propertyA": { "$ref":"./some-non-existent-location.json#definitions/PropertyX"}
+        // validateObject will succeed, because all it cares about is that "$ref" is a URI
+        // In order to validate that $ref points at an existing location in an existing document
+        // we have to "load" the schema
+        loadResourceSchema(definition);
     }
```

Comment:
is there benefit in waiting to validate the object until after the references are loaded, say the customer references something for an object that requires one property, but the referenced object doesnt have it?

Reply:
when i was playing around with this, some constraints in the meta schema weren't being validated when the schema was loaded. conversely, because of the `$ref` is type string issue, validating those wasn't happening in the first step. put together, hopefully this gets both. it's of course possible to ref something that isn't valid, especially since right now, refs aren't restricted. in future, they should be restricted to same-document or relative refs, so then we always know they're referring to another hosted schema (and regionalisation works when we inject `$id`). based on that constraint, it's safe to assume the other schemas are valid, so hopefully there's no issue.

of course, deeply validating the schema would be ideal. but this is already an improvement.

Reply:
makes sense definitely something we can add down the line